### PR TITLE
Form defaults and validation

### PR DESCRIPTION
**Pull request by mishaschwartz**
- use the `jsonschema` library to add defaults to form data before validation

### CODE REVIEW

**Comment by mishaschwartz** [comment]
should be tested with: https://github.com/MarkUsProject/Markus/pull/3954

**Comment by mishaschwartz** [comment]
@david-yz-liu You might notice that the form_validation library still runs the validation twice in some cases.  This is because we can't guarantee that the defaults will be filled in before the `required`, `minItems`, `maxItems`, etc. validators are run.  

One option would be to write a fancy override for each of these validators separately that looks one level down and validates based on what **will be** filled in later on.  This may be slightly more efficient but it would be much more complex to implement and it would introduce opportunities for bugs to creep in. 

I went for option two which fills in **all** the defaults first so we don't have to worry about writing even more custom validators. I did a quick time test and even with our most complex schema, and filling in all defaults possible, it still only takes about 3ms to fully fill in the form and validate. Given that, my vote is to stick with option two.

File: server/form_validation/__init__.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    validate_props = validator_class.VALIDATORS["properties"]
+    validate_array = validator_class.VALIDATORS["items"]
+    validate_oneOf = validator_class.VALIDATORS["oneOf"] # must use draft 4+
+
+    def set_defaults(validator, properties, instance, schema):
+        """ Set defaults within a "properties" context """
+        if not validator.is_type(instance, "object"):
+            return
+        for prop, subschema in properties.items():
+            if instance.get(prop) == None:
```
Comment (david-yz-liu) [inline server/form_validation/__init__.py]: Use `is` instead of `==` for checks against `None`.

File: server/form_validation/__init__.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    """
+    Return an iterator that yields errors from validating obj on schema 
+    after first filling in defaults on obj.
+    """
+    validator = extend_with_default(validator_class)(schema)
+    # first time to fill in defaults since validating 'required', 'minProperties',
+    # etc. can't be done until the instance has been properly filled with defaults.
+    list(validator.iter_errors(obj))
+    return validator.iter_errors(obj)
+
```
Comment (david-yz-liu) [inline server/form_validation/__init__.py]: Remove the extra blank lines (and any trailing whitespace).

File: server/form_validation/__init__.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            msg = f'{instance} is not valid under any of the given schemas'
+            yield ValidationError(msg, context=all_errors)
+        elif len(good_properties) > 1:
+            msg = f'{instance} is valid under each of {", ".join(repr(p) for p in good_properties)}'
+            yield ValidationError(msg)
+        else:
+            instance.clear()
+            instance.update(good_instance)
+
+    custom_validators = {"properties" : set_defaults,
```
Comment (david-yz-liu) [inline server/form_validation/__init__.py]: Remove the space before the colon.

File: server/form_validation/__init__.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    def set_defaults(validator, properties, instance, schema):
+        """ Set defaults within a "properties" context """
+        if not validator.is_type(instance, "object"):
+            return
+        for prop, subschema in properties.items():
+            if instance.get(prop) == None:
+                if "default" in subschema:
+                    default_val = subschema["default"]
+                elif subschema.get('type') == 'array':
```
Comment (david-yz-liu) [inline server/form_validation/__init__.py]: Use one kind of quote consistently. I don't care which, but it seems that double quotes are more common.

**Pull request merged**